src/fileActions.jsx

```diff
@@ -122,14 +124,24 @@ export const ConfirmDeletionDialog = ({
     );
 };
 
-const CreateDirectoryModal = ({ currentPath }) => {
+const CreateDirectoryModal = () => {
     const Dialogs = useDialogs();
     const [name, setName] = useState("");
     const [nameError, setNameError] = useState(null);
     const [errorMessage, setErrorMessage] = useState(undefined);
-    const createDirectory = () => {
-        const path = currentPath + name;
-        cockpit.spawn(["mkdir", path], { superuser: "try", err: "message" })
+    const { currentPath, rootInfo } = useContext(FilesContext);
+
+    const createDirectory = async () => {
+        cockpit.assert(rootInfo, "rootInfo not initialised");
+        const user = await cockpit.user();
+        const options = { err: "message" };
+
+        // Current user does not "own" the current directory so we require superuser
+        if (user.name !== rootInfo.user || user.name !== rootInfo.group) {
+            options.superuser = "try";
```

Comment:
I'm not a fan of this approach, as it still creates files/dirs as root if you are in e.g. /home/someotheruser/ or /var/lib/postgresql/. IMHO a better default is to always create new files/dirs with the same user+group as the current directory. That requires some `runuser` wrapping with cockpit.spawn commands, but that feels not too hard.

Reply:
I'm not sure I like this — it kinda flies in the face of how UNIX is supposed to work.  That's what the `u+s` and `g+s` bits on directories are for (and we should definitely respect them if they're set).

I feel like the correct behaviour is something like "do it as the unprivileged user if it works, otherwise do it as root" with no extra chmod/chown/etc. to make sure things happen in the most "natural" way possible.

Reply:
We won't be using `chown/chmod` either way but run it as the "intended" user ie:
```
runuser -u postgresql -- mkdir foo
```

Doing some try and error research, you need to own the directory to be able to make a sub directory or create a file. So checking the group is wrong and not needed.

So that leaves a decision to be made about what ownership the created file ends up if the user is not the owner of the directory.

Reply:
Nack on creating dirs/files as root in non-root owned directories. Doing that in someone's home directory is weird and wrong. If you don't like `runuser`, then I suppose the compromise is to not allow creating files or dirs in directories which aren't owned by `cockpit.user` or root? Can you live with that?

Reply:
Let's make a bunch of test cases which we should support and then reason about that so:

* Folder with `root:admin` but lacks `g+wr`
* Folder with `root:users` with `g+wr` user should be able to write
* User can create files because of ACL's

Last two should work as normal user, but what happens as superuser.

Reply:
I've revised the whole patch, dropped all the state shenanigans and simplified it.

I come to the conclusion that we have 3 (or more) options:

1) always create as root and chown to parent
2) try to mkdir, on failure use root and chown
2) run test -w and then sudo / chown if needed

As `runuser` is not a solution if we expect a created directory in /var/lib/mock (root:mock) to be created as `root:mock` then we can't use runuser. Along with other cases where mkdir might fail as normal user but are possible as root like a shared dir:

```
drwxr-xr-x. 1 root users 0 May 16 09:28 shared
```